### PR TITLE
Default to STACK_OVERFLOW_CHECK=2 in debug builds

### PR DESCRIPTION
**Pull request by sbc100**
Previously we defaulted to 1 in this case, but we want to improve the error reporting around stack overflow in preparation for lowering the default stack size.  See #14177

Benchmark results:

```
                     Node.js_no_stack_check	Node.js_with_stack_check
              base64 1.000	0.972
        conditionals 1.000	1.096
                copy 1.000	1.003
         corrections 1.000	0.982
       corrections64 1.000	1.004
            fannkuch 1.000	1.004
         fasta_float 1.000	1.002
              havlak 1.000	1.025
                 ifs 1.000	0.962
     matrix_multiply 1.000	1.021
              memops 1.000	0.988
              primes 1.000	0.914
      primes_nocheck 1.000	1.113
            skinning 1.000	0.914
           zzz_box2d 1.000	1.076
          zzz_bullet 1.000	0.988
        zzz_coremark 1.000	0.981
         zzz_linpack 1.000	1.001
 zzz_lua_binarytrees 1.000	1.049
     zzz_lua_scimark 1.000	1.060
            zzz_lzma 1.000	0.984
         zzz_poppler 1.000	1.047
          zzz_sqlite 1.000	1.026
            zzz_zlib 1.000	0.996
```

### CODE REVIEW

**Comment by kripken** [comment]
Are the benchmarks comparing `-sASSERTIONS` before and after this PR (with the benchmarks modified to enable assertions), or do they compare `-O3 -sSTACK_OVERFLOW_CHECK=1` vs ``-O3 -sSTACK_OVERFLOW_CHECK=2`, or something else?

File: test/test_core.py [hunk 1]
```diff
@@ -4131,6 +4131,7 @@ def test_dylink_basics_no_modify(self):
     if self.get_setting('MEMORY64') == 2:
       self.skipTest('MEMORY64=2 always requires module re-writing')
     self.set_setting('WASM_BIGINT')
+    self.set_setting('STACK_OVERFLOW_CHECK', 1)
```
Comment (kripken) [inline test/test_core.py:4134]: Are these just to add more test coverage? Or is it necessary for some reason? Please add a comment.

File: test/test_core.py [hunk 2]
```diff
@@ -8324,6 +8325,7 @@ def test_asyncify_main_module(self):
   def test_emscripten_lazy_load_code(self, conditional):
     if self.get_setting('STACK_OVERFLOW_CHECK'):
       self.skipTest('https://github.com/emscripten-core/emscripten/issues/16828')
+    self.set_setting('STACK_OVERFLOW_CHECK', 0)
```
Comment (kripken) [inline test/test_core.py:8328]: (here too)

File: test/test_other.py [hunk 3]
```diff
@@ -10612,8 +10612,9 @@ def ok(args, filename='hello_world.cpp', expected='hello, world!'):
       self.run_process([EMCC, test_file(filename)] + args)
       self.assertContained(expected, self.run_js('a.out.js'))
 
-    # -O0 with BigInt support (to avoid the need for legalization)
-    required_flags = ['-sWASM_BIGINT']
+    # -O0 with BigInt support (to avoid the need for legalization) and STACK_OVERFLOW_CHECK=1 (to
+    # avoid the need for the stack check pass).
+    required_flags = ['-sWASM_BIGINT', '-sSTACK_OVERFLOW_CHECK=1']
```
Comment (kripken) [inline test/test_other.py:10617]: Hmm, this means the default build will still do wasm changes after link even after we default to wasm-bigint. Maybe it makes sense though.

Please add to this test. We want to emit a good error message in this case, that is, `ERROR_ON_WASM_CHANGES_AFTER_LINK` should do like it does with `WASM_BIGINT` - if it's not set right, suggest the proper value.

**Review by kripken — COMMENTED** [review]
I think this probably makes sense to do, since performance in debug builds isn't critical, and it looks like this is not a significant regression there anyhow.

**Comment by sbc100** [comment]
> Are the benchmarks comparing `-sASSERTIONS` before and after this PR (with the benchmarks modified to enable assertions), or do they compare `-O3 -sSTACK_OVERFLOW_CHECK=1` vs `-O3 -sSTACK_OVERFLOW_CHECK=2`, or something else?

These benchmarks are actually from `main` I ran with:

```
a: ['-sASSERTIONS', '-sSTACK_OVERFLOW_CHECK=1']
b: ['-sASSERTIONS', '-sSTACK_OVERFLOW_CHECK=2']
```

File: test/test_core.py (hunk 2 above)
Comment (sbc100) [reply test/test_core.py:8328]: This is because LAZY_LOAD_CODE is broken with STACK_OVERFLOW_CHECK.. see the bug on the line above.

**Comment by kripken** [comment]
Are those flags on top of `-O3` or something else? Or at those all the flags?

**Comment by sbc100** [comment]
> Are those flags on top of `-O3` or something else? Or are those all the flags passed to emcc?

I think they are on top of `-O3`. 

`EmscriptenBenchmarker('Node.js no stack check', config.NODE_JS, ['-sASSERTIONS', '-sSTACK_OVERFLOW_CHECK=1'])`

I think the EmscriptenBenchmarker injects `OPTIMIZATIONS` in addition to these flags, right?

**Comment by kripken** [comment]
Yes, that sounds right.

**Comment by sbc100** [comment]
Closing a favor of #18154.   After speaking with @dschuff we decided we really didn't want to add any binaryen passed to debug builds by default since that would inhibit that ability to skip binaryen rewriting completely (which is very important for fast debug link times)

**Pull request closed without merging**